src/densities.jl

```diff
@@ -76,33 +79,42 @@ is not collinear the spin density is `nothing`.
     end
 
     Threads.@threads for (ikpts, ρaccu) in collect(zip(kpt_per_thread, ρaccus))
-        for iσ in 1:n_spin
-            ρaccu[iσ] .= 0
-        end
+        ρaccu .= 0
         ρ_k = similar(ψ[1][:, 1], basis.fft_size)
         for ik in ikpts
             kpt = basis.kpoints[ik]
             compute_partial_density!(ρ_k, basis, kpt, ψ[ik], occupation[ik])
             lowpass_for_symmetry!(ρ_k, basis)
             # accumulates all the symops of ρ_k into ρaccu
-            accumulate_over_symmetries!(ρaccu[kpt.spin], ρ_k, basis, basis.ksymops[ik])
+            accumulate_over_symmetries!(ρaccu[:, :, :, kpt.spin], ρ_k, basis, basis.ksymops[ik])
         end
     end
 
     # Count the number of k-points modulo spin
     count = sum(length(basis.ksymops[ik]) for ik in 1:length(basis.kpoints)) ÷ n_spin
     count = mpi_sum(count, basis.comm_kpts)
-    ρs = [sum(getindex.(ρaccus, iσ)) / count for iσ in 1:n_spin]
-    mpi_sum!.(ρs, Ref(basis.comm_kpts))
+    ρ = sum(ρaccus) ./ count
+    mpi_sum!(ρ, basis.comm_kpts)
+    G_to_r(basis, ρ)
+end
 
-    @assert basis.model.spin_polarization in (:none, :spinless, :collinear)
-    if basis.model.spin_polarization == :collinear
-        ρtot  = from_fourier(basis, ρs[1] + ρs[2])  # up + down
-        ρspin = from_fourier(basis, ρs[1] - ρs[2])  # up - down
+total_density(ρ) = dropdims(sum(ρ; dims=4); dims=4)
+@views function spin_density(ρ)
```

Comment:
These are not just useful for the density, but also for the potential, so it would be nice to see that reflected in the names, so how about `compute_total` and `compute_spin`? Also almost always you need both of them if you need one. Maybe we should just have a function `split_total_and_spin` or `compute_total_and_spin` to reflect that?

Also at least in theory these definitions change once we go beyond collinear spin ... I don't think we should worry about that now, but just to keep that in mind. That's amongst the reasons why I would go beyond a plain array, but instead a special type for spin-component arrays. Then you can just put a reference to a basis in the SpinArray and define total(::SpinArray) and spin(::SpinArray) and all is well. But of course we can still do that later when we need it.

Reply:
Yes, but we manipulate densities in this 4D format a lot more than potentials, at least for now, so I kept it density specific for clarity. Yes I know non-collinear spin doesn't fit this API, but at least as a first non-optimized version we can do 5D (complex) densities [i,j,k,alpha,beta]: let's keep things simple for now and we'll see when we get to it.

Reply:
Ah you only refactored the densities and not the potentials as well. I have not read enough of the code to see that. But still ... perhaps go with the general names. What happens should be clear from context (we pretty much always use `ρ` or `V` as part of the variable names).